src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

The implementation of the `WebhookSecretGuard` looks good and handles webhook signature verification. However, there are a few improvements that can be suggested:

1. Consider adding error handling for potential exceptions thrown by `crypto.createHmac`.
2. Ensure proper validation of `Env.SECRET.WEBHOOK` before using it.
3. Add comments to explain the purpose of each section of the code for better readability.

```diff
+     try {
+         const hmac = crypto.createHmac('sha256', secret);
+         const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+     } catch (error) {
+         throw new UnauthorizedException('Error creating HMAC');
+     }
+
+     // Validate Env.SECRET.WEBHOOK
+
```

Overall, the guard implementation is secure and follows best practices for verifying webhook signatures.



<!-- This is an auto-generated comment by PullProofAI -->